woocommerce/woocommerce
# [Accessibility] Announce notice messages after page loads

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:




1. This PR introduces a JS code that looks for any notice element with the `role="alert"` attribute on the page and focuses it once the page loads.
2. The markup of the success message of the add-to-cart action has been changed. There were two accessibility issues with the old markup:
    1. The visual order of the elements didn't match the DOM order, so the focus changing didn't follow a logical order.
    2. Once the notice message was focused, the link's content was announced before the message itself.

#### Before
The link comes before the message.
![image](https://github.com/user-attachments/assets/9c6f522e-656d-4aab-9da4-a5f688226bbe)
![image](https://github.com/user-attachments/assets/087e9793-0e65-4ad0-b26f-ff465ea56e0d)
![image](https://github.com/user-attachments/assets/d7492814-d615-4c49-a99e-b2ecae6917af)


#### After
The link comes after the message.
![image](https://github.com/user-attachments/assets/e11a7678-001f-440b-a2eb-458c8aded760)
![notice-add-to-cart](https://github.com/user-attachments/assets/cd818e5e-2309-4ce4-9ef4-f4b4534e5375)
![notice-stock](https://github.com/user-attachments/assets/e7d0937e-2e00-464a-bbf5-137752c63217)


The change in the markup doesn't entail a significant change in the visual appearance. In fact, two other notice messages had already the link coming after the text. They are:
- My Account -> Downloads
![notice-no-downloads](https://github.com/user-attachments/assets/dc61a825-d81f-41d0-9944-96ed42c9c165)
- My Account -> Order
![notice-no-orders](https://github.com/user-attachments/assets/a8c7f98f-30f1-44bf-bd80-7dff0b1f7840)


Closes #43614 .
Closes #51931 .
Closes #51929 .
Closes #52142 .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Testing the add-to-cart notice
1. Install the Twenty Twenty-Four theme.
2. Go to a Product page.
3. Add the product to the cart.
4. Verify the error message gets the focus once the page loads.


#### Testing the stock notice
1. Edit the stock quantity of a product to have only one unit available.
2. Go to the product page on the frontend.
3. Add two products to the cart.
4. Verify the error message gets the focus once the page loads.

#### Testing the Address form notices
1. On the frontend, go to My Account -> Addresses -> Edit address.
2. Leave a required field empty and submit the form.
3. Verify the error message gets the focus once the page loads.




### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
